mender-update/deployments.cpp

```diff
@@ -0,0 +1,207 @@
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <sstream>
+#include <string>
+
+#include <common/common.hpp>
+#include <common/error.hpp>
+#include <common/events.hpp>
+#include <common/expected.hpp>
+#include <common/http.hpp>
+#include <common/io.hpp>
+#include <common/json.hpp>
+#include <common/log.hpp>
+#include <common/optional.hpp>
+#include <common/path.hpp>
+#include <mender-update/context.hpp>
+
+#include <mender-update/deployments.hpp>
+
+namespace mender {
+namespace update {
+namespace deployments {
+
+using namespace std;
+
+namespace common = mender::common;
+namespace context = mender::update::context;
+namespace error = mender::common::error;
+namespace events = mender::common::events;
+namespace expected = mender::common::expected;
+namespace io = mender::common::io;
+namespace json = mender::common::json;
+namespace log = mender::common::log;
+namespace optional = mender::common::optional;
+namespace path = mender::common::path;
+
+const DeploymentsErrorCategoryClass DeploymentsErrorCategory;
+
+const char *DeploymentsErrorCategoryClass::name() const noexcept {
+	return "DeploymentsErrorCategory";
+}
+
+string DeploymentsErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	case InvalidDataError:
+		return "Invalid data error";
+	case BadResponseError:
+		return "Bad response error";
+	}
+	assert(false);
+	return "Unknown";
+}
+
+error::Error MakeError(DeploymentsErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, DeploymentsErrorCategory), msg);
+}
+
+const string v1_uri = "api/devices/v1/deployments/device/deployments/next";
+const string v2_uri = "api/devices/v2/deployments/device/deployments/next";
+
+error::Error CheckNewDeployments(
+	context::MenderContext &ctx,
+	const string &server_url,
+	http::Client &client,
+	events::EventLoop &loop,
+	APIResponseHandler api_handler) {
+	auto ex_dev_type = ctx.GetDeviceType();
+	if (!ex_dev_type) {
+		return ex_dev_type.error();
+	}
+	string device_type = ex_dev_type.value();
+
+	auto ex_provides = ctx.LoadProvides();
+	if (!ex_provides) {
+		return ex_provides.error();
+	}
+	auto provides = ex_provides.value();
+	if (provides.find("artifact_name") == provides.end()) {
+		return MakeError(InvalidDataError, "Missing artifact name data");
+	}
+	string artifact_name = provides["artifact_name"];
+
+	stringstream ss;
+	ss << R"({"update_control_map": false, "device_provides": {)";
+	ss << R"("device_type": ")";
+	ss << json::EscapeString(device_type);
```

Comment:
Q: Are we not missing some device provides here?

Reply:
@kacf should this add the other pairs from `LoadProvides()`? The docs 
![image](https://github.com/mendersoftware/mender/assets/1044278/47cfec34-ea30-42db-abcf-6fb5d7f1fbbd)
are not super helpful for me, no clue what `additionalProperties` actually means.

Reply:
Just for the record: the answer is yes.